File: agent/linux/pkg/selftest/SelfTestRunner.go
package selftest

import (
	"fmt"
	"net/http"
	"os"
	"time"
)

type TestResult struct {
	Name    string `json:"name"`
	Passed  bool   `json:"passed"`
	Message string `json:"message,omitempty"`
}

type SelfTestRunner struct {
	backendURL string
}

func NewSelfTestRunner(backendURL string) *SelfTestRunner {
	return &SelfTestRunner{backendURL: backendURL}
}

// RunAll executes all health checks and returns results.
func (s *SelfTestRunner) RunAll() []TestResult {
	tests := []struct {
		name string
		fn   func() (bool, string)
	}{
		{"backend_reachable", s.testBackend},
		{"running_as_root", s.testRoot},
		{"quarantine_dir_writable", s.testQuarantineDir},
		{"evidence_dir_writable", s.testEvidenceDir},
	}

	var results []TestResult
	for _, t := range tests {
		passed, msg := t.fn()
		results = append(results, TestResult{Name: t.name, Passed: passed, Message: msg})
		if passed {
			fmt.Printf("[SelfTest] ✅ %s\n", t.name)
		} else {
			fmt.Printf("[SelfTest] ❌ %s: %s\n", t.name, msg)
		}
	}
	return results
}

func (s *SelfTestRunner) testBackend() (bool, string) {
	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Get(s.backendURL)
	if err != nil {
		return false, fmt.Sprintf("backend unreachable: %v", err)
	}
	resp.Body.Close()
	return true, ""
}

func (s *SelfTestRunner) testRoot() (bool, string) {
	if os.Getuid() != 0 {
		return false, "agent must run as root for eBPF"
	}
	return true, ""
}

func (s *SelfTestRunner) testQuarantineDir() (bool, string) {
	return testDirWritable("/var/lib/fenrir/quarantine")
}

func (s *SelfTestRunner) testEvidenceDir() (bool, string) {
	return testDirWritable("/var/lib/fenrir/evidence")
}

func testDirWritable(path string) (bool, string) {
	if err := os.MkdirAll(path, 0700); err != nil {
		return false, fmt.Sprintf("cannot create %s: %v", path, err)
	}
	tmp := path + "/.fenrir_write_test"
	if err := os.WriteFile(tmp, []byte("ok"), 0600); err != nil {
		return false, fmt.Sprintf("not writable: %v", err)
	}
	os.Remove(tmp)
	return true, ""
}

File: agent/macos/Sources/FenrirAgent/SelfTestRunner.swift
import Foundation

struct TestResult {
    let name: String
    let passed: Bool
    let message: String?
}

class SelfTestRunner {
    private let config: AgentConfig

    init(config: AgentConfig) {
        self.config = config
    }

    func runAll() -> [TestResult] {
        let tests: [(name: String, fn: () -> (Bool, String?))] = [
            ("backend_reachable", testBackend),
            ("running_as_root", testRoot),
            ("quarantine_dir_writable", testQuarantineDir),
            ("evidence_dir_writable", testEvidenceDir),
            ("full_disk_access", testFullDiskAccess)
        ]

        var results: [TestResult] = []
        for test in tests {
            let (passed, message) = test.fn()
            results.append(TestResult(name: test.name, passed: passed, message: message))
            let icon = passed ? "✅" : "❌"
            print("[SelfTest] \(icon) \(test.name)\(message.map { ": \($0)" } ?? "")")
        }
        return results
    }

    private func testBackend() -> (Bool, String?) {
        let semaphore = DispatchSemaphore(value: 0)
        var success = false
        let url = config.backendURL.deletingLastPathComponent()
        URLSession.shared.dataTask(with: url) { _, response, _ in
            success = (response as? HTTPURLResponse) != nil
            semaphore.signal()
        }.resume()
        semaphore.wait()
        return (success, success ? nil : "backend unreachable at \(url)")
    }

    private func testRoot() -> (Bool, String?) {
        let isRoot = getuid() == 0
        return (isRoot, isRoot ? nil : "agent should run as root for full ES entitlement")
    }

    private func testQuarantineDir() -> (Bool, String?) {
        return testWritable("Fenrir/Quarantine")
    }

    private func testEvidenceDir() -> (Bool, String?) {
        return testWritable("Fenrir/Evidence")
    }

    private func testFullDiskAccess() -> (Bool, String?) {
        // Probe a protected path to confirm FDA is granted
        let probe = "/Library/Application Support/com.apple.TCC/TCC.db"
        let accessible = FileManager.default.isReadableFile(atPath: probe)
        return (accessible, accessible ? nil : "Full Disk Access not granted in System Preferences")
    }

    private func testWritable(_ subpath: String) -> (Bool, String?) {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .localDomainMask).first!
        let dir = base.appendingPathComponent(subpath)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let test = dir.appendingPathComponent(".write_test")
        do {
            try "ok".write(to: test, atomically: true, encoding: .utf8)
            try FileManager.default.removeItem(at: test)
            return (true, nil)
        } catch {
            return (false, "not writable: \(error.localizedDescription)")
        }
    }
}
